lib/journaled/abstract_adapter_ext.rb

```diff
@@ -0,0 +1,68 @@
+module Journaled
+  module AbstractAdapterExt
+    def self.included(klass)
+      klass.set_callback(:checkout, :before) do
+        @_journaled_transaction_handler = nil
+      end
+    end
+
+    def _journaled_transaction_handler
+      if @_journaled_transaction_handler&.active?
+        @_journaled_transaction_handler
+      else
+        @_journaled_transaction_handler = TransactionHandler.new(connection: self)
+      end
+    end
```

Comment:
This `AbstractAdapterExt` extends Rails' abstract connection adapter so that we can set up and manage a bit of thread-safe context. I'm doing the actual `include` [over here](https://github.com/Betterment/journaled/pull/25/files#diff-28abb7401d95adc813455700fc0d2c247c4a3487553660e19407e6aa02fe6008R10) inside of `engine.rb`.

Reply:
I think we're going to need transaction handlers per transaction, rather than a single one on the connection (or something similar). The case where this comes up is if you have nested (savepoint) transaction, where the inner transaction does a rollback. With a single handler, if there was already an event batched in the outer transaction, then events emitted in the inner one are added to the handler attached to the outer transaction. If the inner transaction rolls back, those events are still sent when the outer transaction commits.

An example can be found in this [spec](https://github.com/smudge/journaled/compare/event-batching...medlefsen:journaled:me-event-batching#diff-d150802de35523c1e68d9701f8668a124dcfb9fc5c9e23c197442b7c449e653aR286-R298).

I starting thinking through a way around this using the transaction records, but the other issue with before_commit hooks makes me think it likely needs to be seperate from the transaction records system (but work similarly in terms of savepoints and joinable transactions).